xapian-applications/omega/omegatest

```diff
@@ -655,6 +655,28 @@ testcase 'b4c216e4da73d1d01277ef46d0514821' text="simple query"
 testcase 'd41d8cd98f00b204e9800998ecf8427e' text=""
 testcase '3302c94d3500c3f695b7e09b7acd420d' text=" test "
 
+# Feature tests for random{}
+rng_test_failed=0
+rng_test() {
+    expected=$1
+    shift
+    [ "$#" != 0 ] || { set x dummy; shift; }
+    output=`$FAKETIME ${FAKE_NOW+"$FAKE_NOW"} "$OMEGA" "$@"`
+    if [ "$output" -gt "$expected" ] || [ "$output" -lt 0 ]; then
+	echo "$OMEGA $@:"
+	printf '%s\n' "  expected: «${expected}»"
+	printf '%s\n' "  received: «${output}»"
+	if [ "$rng_test_failed" -eq 0 ]; then
+            failed=`expr $failed + 1`
+            rng_test_failed=1
+        fi
+    fi
+}
+for i in `seq 100`; do
+    printf '$random{$cgi{x}}' > "$TEST_TEMPLATE"
+    rng_test '6' x=6
+done
```

Comment:
While this looks functionally correct to me, it runs omega 100 times and adding a bit of timing code around it shows that it takes just over 2 seconds out of the ~12 seconds that omegatest now takes to run on my dev machine - that's a lot of overhead for testing a single feature.

We should be able to perform all the testing in one run of omega - `$map{$range{1,100},X}` will execute `X` 100 times and give us a list with all the results, so we can make `X` generate a single random value, range-test it, and give an empty string if all is well and a non-empty string otherwise, then we can concatenate all the entries in the resulting list (with `$list{MAPCOMMAND,}`) and the result will be empty if all 100 trials were good and non-empty otherwise (the `$or` below will give us "true" for each trial which failed the range check).

Putting that together we get something like this, which should output "OK" if the test passed (if an exception is thrown or omega crashes we'll get no output, so I've deliberately avoided having empty output mean "success"):

```
$and{$list{$map{$range{1,100},$set{num,$random{$cgi{x}}}$or{$lt{$opt{num},0},$gt{$opt{num},$cgi{x}}}},},OK}
```

And that takes about 0.022 seconds to run, reducing omegatest's runtime to about 10 seconds.

Don't worry about applying that - I'll merge your branch and add in that change as well shortly.

Reply:
> Don't worry about applying that - I'll merge your branch and add in that change as well shortly.

Oh, except I then spotted some other minor niggles.